src/cpp/core/Database.cpp

```diff
@@ -0,0 +1,635 @@
+   };
+   std::sort(files.begin(), files.end(), comparator);
+
+   *pVersion = files.at(0).getStem();
+   return Success();
+}
+
+Error SchemaUpdater::isSchemaVersionPresent(bool* pIsPresent)
+{
+   std::string queryStr;
+   if (connection_->driverName() == SQLITE_DRIVER)
+   {
+      queryStr = std::string("SELECT COUNT(1) FROM sqlite_master WHERE type='table' AND name='") + SCHEMA_TABLE + "'";
+   }
+   else if (connection_->driverName() == POSTGRESQL_DRIVER)
+   {
+      queryStr = std::string("SELECT COUNT(1) FROM information_schema.tables WHERE table_name='") + SCHEMA_TABLE +
+                 "' AND table_schema = current_schema";
+   }
+   else
+   {
+      return DatabaseError(soci::soci_error("Unsupported database driver"));
+   }
+
+   int count = 0;
+   Query query = connection_->query(queryStr)
+         .withOutput(count);
+   Error error = connection_->execute(query);
+   if (error)
+      return error;
+
+   *pIsPresent = count > 0;
+   return Success();
+}
+
+Error SchemaUpdater::databaseSchemaVersion(std::string* pVersion)
+{
+   bool versionPresent = false;
+   Error error = isSchemaVersionPresent(&versionPresent);
+   if (error)
+      return error;
+
+   std::string currentSchemaVersion = "0";
+   if (!versionPresent)
+   {
+      // no schema version present - add the table to the database so it is available
+      // for updating whenever migrations occur
+      error = connection_->executeStr(std::string("CREATE TABLE ") + SCHEMA_TABLE + "(current_version text)");
+      if (error)
+         return error;
+
+      Query query = connection_->query(std::string("INSERT INTO ") + SCHEMA_TABLE + " VALUES (:val)")
+            .withInput(currentSchemaVersion);
+      error = connection_->execute(query);
+      if (error)
+         return error;
+
+      *pVersion = currentSchemaVersion;
+      return Success();
+   }
+
+   Query query = connection_->query(std::string("SELECT current_version FROM ") + SCHEMA_TABLE)
+         .withOutput(currentSchemaVersion);
+
+   error = connection_->execute(query);
+   if (error)
+      return error;
+
+   *pVersion = currentSchemaVersion;
+   return Success();
+}
+
+Error SchemaUpdater::isUpToDate(bool* pUpToDate)
+{
+   std::string version;
+   Error error = databaseSchemaVersion(&version);
+   if (error)
+      return error;
+
+   std::string migrationVersion;
+   error = highestMigrationVersion(&migrationVersion);
+   if (error)
+      return error;
+
+   *pUpToDate = version >= migrationVersion;
+   return Success();
+}
+
+Error SchemaUpdater::update()
+{
+   std::string migrationVersion;
+   Error error = highestMigrationVersion(&migrationVersion);
+   if (error)
+      return error;
+
+   std::string currentVersion;
+   error = databaseSchemaVersion(&currentVersion);
+   if (currentVersion < migrationVersion)
+      return updateToVersion(migrationVersion);
+   else
+      return Success();
+}
+
+Error SchemaUpdater::updateToVersion(const std::string& maxVersion)
```

Comment:
I'm not as familiar on what the best practices are re: database updates + schema versioning. Do you have any recommended resources for getting better acquainted?

Reply:
This was pretty much taken from Connect's implementation: https://github.com/rstudio/connect/blob/master/src/connect/store/version.go#L261-L345

This is similar to how many popular tools do things as well. See https://flywaydb.org/documentation/migrations